Sources/MongoSwift/APM.swift

```diff
@@ -14,6 +14,133 @@ public struct ConnectionId {
+    /// record this field when monitoring config servers as replica sets.
+    let opTime: ObjectId? = nil
+
+    /// The type of this server.
+    let type: ServerType = .unknown
+
+    /// The wire protocol version range supported by the server.
+    let minWireVersion: Int32 = 0
+    let maxWireVersion: Int32 = 0
+
+    /// The hostname or IP and the port number that this server was configured with in the replica set.
+    let me: ConnectionId? = nil
+
+    /// Hosts, arbiters, passives: sets of addresses. This server's opinion of the replica set's members, if any.
+    let hosts: [ConnectionId] = []
+    let arbiters: [ConnectionId] = []
+    /// "Passives" are priority-zero replica set members that cannot become primary. 
+    /// The client treats them precisely the same as other members.
+    let passives: [ConnectionId] = []
+
+    /// Tags for this server.
+    let tags: [String: String] = [:]
+
+    /// The replica set name.
+    let setName: String? = nil
+
+    /// The replica set version.
+    let setVersion: Int64? = nil
+
+    /// The election ID where this server was elected, if this is a replica set member that believes it is primary.
+    let electionId: ObjectId? = nil
+
+    /// This server's opinion of who the primary is. 
+    let primary: ConnectionId? = nil
+
+    /// When this server was last checked.
+    let lastUpdateTime: Date? = nil
+
+    /// The logicalSessionTimeoutMinutes value for this server.
+    let logicalSessionTimeoutMinutes: Int64? = nil
+}
+
+/// The possible types for a topology. The raw values correspond to the values libmongoc uses. 
+/// (We don't use these strings directly because Swift convention is to use lowercase for enums.)
+public enum TopologyType: String {
+    case single = "Single"
+    case replicaSetNoPrimary = "ReplicaSetNoPrimary"
+    case replicaSetWithPrimary = "ReplicaSetWithPrimary"
+    case sharded = "Sharded"
+    case unknown = "Unknown"
+}
+
+/// A struct describing the state of a MongoDB deployment: its type (standalone, replica set, or sharded), 
+/// which servers are up, what type of servers they are, which is primary, and so on.
+public struct TopologyDescription {
+    /// The type of this topology. 
+    let type: TopologyType
+
+    /// The replica set name. 
+    let setName: String? = nil
+
+    /// The largest setVersion ever reported by a primary.
+    let maxSetVersion: Int64? = nil
+
+    /// The largest electionId ever reported by a primary.
+    let maxElectionId: ObjectId? = nil
+
+    /// The servers comprising this topology. By default, a single server at localhost:270107.
+    let servers: [ServerDescription] = [ServerDescription(connectionId: ConnectionId())]
+
+    /// For single-threaded clients, indicates whether the topology must be re-scanned.
+    let stale: Bool = false
+
+    /// Exists if any server's wire protocol version range is incompatible with the client's.
+    let compatibilityError: MongoError? = nil
```

Comment:
they suggest a bool indicating an error and then a second string property containing the error message, but this seems cleaner to me